### ISSUE TITLE
Exception when disabling watchdog

### ISSUE THREAD

**Issue by ctron** [eclipse/kura#2010]
Disabling the watchdog causes the following exception in the Kura log:

~~~
2018-04-03 15:37:46,002 [Component Resolve Thread (Bundle 7)] INFO  o.e.k.l.w.WatchdogServiceImpl - WatchdogServiceImpl task cancelled? = true
2018-04-03 15:37:46,006 [WatchdogServiceImpl] ERROR o.e.k.l.w.WatchdogServiceImpl - Failed to perform watchdog actions.
org.eclipse.kura.KuraException: Command ''sync' failed' exited with code {1}.
        at org.eclipse.kura.linux.watchdog.WatchdogServiceImpl.runCommand(WatchdogServiceImpl.java:222)
        at org.eclipse.kura.linux.watchdog.WatchdogServiceImpl.performWatchdogActions(WatchdogServiceImpl.java:210)
        at org.eclipse.kura.linux.watchdog.WatchdogServiceImpl.checkCriticalComponents(WatchdogServiceImpl.java:193)
        at org.eclipse.kura.linux.watchdog.WatchdogServiceImpl.lambda$0(WatchdogServiceImpl.java:88)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.InterruptedException
        at java.lang.Object.wait(Native Method)
        at java.lang.Object.wait(Object.java:502)
        at java.lang.UNIXProcess.waitFor(UNIXProcess.java:395)
        at org.eclipse.kura.linux.watchdog.WatchdogServiceImpl.runCommand(WatchdogServiceImpl.java:220)
        ... 10 more

~~~

**Comment by MMaiero** [comment]
Never seen that in other systems. Is that causing issues?

**Comment by cdealti** [comment]
For sure the KuraException is not properly built.

The format message is intended to produce something like `Command 'sync' failed with code -1` but here the wait for the command `sync` to complete is interrupted so there's no exit code.
An example is [here](https://github.com/eclipse/kura/blob/release-3.2.0/kura/org.eclipse.kura.linux.net/src/main/java/org/eclipse/kura/linux/net/util/IpAddrShow.java#L67).

@ctron It would be useful to have a little more context (log lines before the exception) and an explanation on how you managed to disable the watchdog at the precise time the watchdog service was executing the sync+reboot action.

@MMaiero 
> Never seen that in other systems. Is that causing issues?

I don't think it's causing issues. The sync+reboot action is simply aborted as expected if the service is disabled. However let's wait for a confirm by @ctron.

**Comment by ctron** [comment]
As there are other issues with the watchdog, I am not sure this didn't have any impact. So short answer: I don't know.

The full set of logs is here: [kura-logs-20180404-090416.tar.gz](https://github.com/eclipse/kura/files/1874733/kura-logs-20180404-090416.tar.gz)

**Comment by ctron** [comment]
So looking at the code shows that this probably wasn't happening when disabling the watchdog, but when the watchdog wanted to reboot:

~~~java
    private synchronized void performWatchdogActions() throws KuraException {
        // watchdogToStop is used to avoid multiple action executions and kill Kura after a while
        if (!this.watchdogToStop) {
            this.watchdogToStop = true;
            runCommand("sync");
            runCommand("reboot");
        } else {
            refreshWatchdog();
        }
    }
~~~

Now if the call to `sync` fails, then the reboot command it not being executed. However the watchdog is disabled at the same time.

I am not sure why the call to `sync` returns a non-zero result, but in any case the reboot should be executed IMHO.

**Comment by ctron** [comment]
Which I think also explains the follow up issues like #2011.

**Comment by ctron** [comment]
Ok, looking more closely the issue gets more complex. 

The exception is not coming from the call to `sync`, but to the combination of modifying the watchdog and a reboot being in progress. The call to sync took a while (as it might happen) and at the same time (not intentionally) I did disable the watchdog (which causes an update). Which then causes a call to `cancel` on the poll task, which was busy waiting for `sync` at the same time. So it is not the `sync` command which failed, but the Java task of waiting for the return. However `IOException` is treated the same as `InterruptedException`.

So I think there are a few things to fix here:

1. A failed call to `sync` should not prevent a reboot
1. The shutdown of the watchdog should not cancel the task with `mayInterruptIfRunning` set to `true`. If the reboot is in progress then it should continue.
1. I think the poll task could be left alone if the ping interval didn't change
1. Also the logic of the `performWatchdogActions` seems a little bit odd to me. Maybe this is again designed that way and I simply don't understand it. But if the watchdog should be stopped (`watchdogToStop = true`) and the call gets executed again (like in this scenario), then the watchdog is refreshed. I do understand that the HW watchdog should be kept alive as long as the "sync" is running. But in the cause either "sync" or "reboot" fail, the watchdog should not be refreshed any longer. So in case of any exception to one of those two calls, refreshing the watchdog should stop IMHO.

**Comment by cdealti** [comment]
> The exception is not coming from the call to sync, but to the combination of modifying the watchdog and a reboot being in progress. The call to sync took a while (as it might happen) and at the same time (not intentionally) I did disable the watchdog (which causes an update).

This is what I was asking in my first comment.

The outcome of this race condition, disabling the `WatchdogService` while it was running the shutdown actions, effectively canceled the shutdown. It does not look a bug.

I agree that the current code may be improved with your suggestions and we will consider them for the next development cycle.

**Comment by cdealti** [comment]
@ctron OK, I should give up with software ;).

If the WatchdogService is re-enabled later on, the damned flag `watchdogToStop` remains set to true (I hate flags), preventing further watchdog actions.

It must be fixed for 3.2.0.

**Comment by MMaiero** [comment]
It seems to be verified on the supported platforms. Let's close for now
